File: unity-package/Editor/StateTree/StateTree.cs
using System.Collections.Generic;
using System;
using System.Linq;
using UnityEngine;
using System.Text;
// Migrated from Newtonsoft.Json to SimpleJson

namespace UnityMcp
{
    /// <summary>
    /// StateTreeExecution context wrapper class，SupportJSONSerialized fields and non-serialized object references
    /// </summary>
    public class StateTree
    {
        public string key;                         // Variable at current layer
        public Dictionary<object, StateTree> select = new();
        public HashSet<string> optionalParams = new(); // Storage for optional parameterskey
        public Func<JsonClass, object> func;     // Leaf function（Backward compatibility，Already from JsonClass Migrate to JSONClass）
        public Func<StateTreeContext, object> contextFunc; // New context leaf function
        public const string Default = "*";          // Wildcard identifier
        public string ErrorMessage;//Execute error message

        /* Implicit conversion：Action → Leaf node（Backward compatibility，Already from JsonClass Migrate to JSONClass） */
        public static implicit operator StateTree(Func<JsonClass, object> a) => new() { func = a };

        /* Implicit conversion：ContextAction → Leaf node（New version） */
        public static implicit operator StateTree(Func<StateTreeContext, object> a) => new() { contextFunc = a };

        /* Run：Along the unique path of the tree（JSONClass Context，Backward compatibility） */
        public object Run(JsonClass ctx, Dictionary<string, object> dict = null)
        {
            // Convert toStateTreeContextAnd call the new version'sRunMethod
            StateTreeContext context = new StateTreeContext(ctx, dict ?? new Dictionary<string, object>());
            return Run(context);
        }

        /* Run：Along the unique path of the tree（StateTreeContext Context） */
        public object Run(StateTreeContext ctx)
        {
            var cur = this;
            while (cur.func == null && cur.contextFunc == null)
            {
                object keyToLookup = Default;
                StateTree next = null;

                // First check if there is a regular (matching)keyMatch
                if (!string.IsNullOrEmpty(cur.key) && ctx != null && ctx.TryGetJsonValue(cur.key, out JsonNode token))
                {
                    keyToLookup = ConvertTokenToKey(token);
                    cur.select.TryGetValue(keyToLookup, out next);
                }

                // If no regular match is found，Check optional parameters
                if (next == null && ctx != null)
                {
                    // Find all optional parameter keys
                    foreach (var kvp in cur.select)
                    {
                        if (kvp.Key == null) continue; // SkipnullKey

                        string key = kvp.Key.ToString();
                        if (!string.IsNullOrEmpty(key) && cur.optionalParams.Contains(key))
                        {
                            // Check if parameter exists and is not empty
                            if (ctx.TryGetJsonValue(key, out JsonNode paramToken) &&
                                paramToken != null &&
                                paramToken.type != JsonNodeType.Null &&
                                !string.IsNullOrEmpty(paramToken.Value))
                            {
                                next = kvp.Value;
                                break; // Use the first matching optional parameter found
                            }
                        }
                    }
                }

                // If still not found，Try default branch
                if (next == null && !cur.select.TryGetValue(Default, out next))
                {
                    var supportedKeys = cur.select.Keys
                        .Where(k => k?.ToString() != Default && !(cur.optionalParams.Contains(k?.ToString())))
                        .Select(k => k?.ToString() ?? "null")
                        .ToList();

                    // Add optional parameter to supported key list
                    var optionalKeys = cur.select.Keys
                        .Where(k => k != null && cur.optionalParams.Contains(k.ToString()))
                        .Select(k => k.ToString() + " (optional)")
                        .ToList();

                    supportedKeys.AddRange(optionalKeys);

                    string supportedKeysList = supportedKeys.Count > 0
                        ? string.Join(", ", supportedKeys)
                        : "none";

                    ErrorMessage = $"Invalid value '{keyToLookup}' for key '{cur.key}'. Supported values: [{supportedKeysList}]";
                    return null;
                }
                cur = next;
            }

            // Prefer to use new versioncontextFunc，If notfunc
            if (cur.contextFunc != null)
            {
                return cur.contextFunc.Invoke(ctx);
            }
            else if (cur.func != null)
            {
                return cur.func.Invoke(ctx?.JsonData);
            }

            return null;
        }

        private static object ConvertTokenToKey(JsonNode token)
        {
            if (token == null || token.type == JsonNodeType.Null)
                return Default;

            if (token.type == JsonNodeType.Integer)
            {
                int longVal = token.AsInt;
                if (longVal <= int.MaxValue && longVal >= int.MinValue)
                {
                    return (int)longVal;
                }
                return longVal;
            }

            if (token.type == JsonNodeType.Float)
            {
                return token.AsDouble;
            }

            if (token.type == JsonNodeType.Boolean)
            {
                return token.AsBool;
            }

            if (token.type == JsonNodeType.String)
            {
                return token.Value;
            }

            // JsonData Type directly returns value
            if (token is JsonData jsonData && !string.IsNullOrEmpty(jsonData.Value))
            {
                return jsonData.Value;
            }

            return token.Value;
        }

        /* Beautified print（Unicode Box line） */
        public void Print(StringBuilder sb, string indent = "", bool last = true, string parentEdgeLabel = null)
        {
            // Root node：Print title
            if (string.IsNullOrEmpty(indent))
            {
                sb.AppendLine($"{indent}StateTree");
            }

            // If current node has key，Print node once key（Avoid duplicating parent edge tag）
            string edgesIndent = indent;
            if (!string.IsNullOrEmpty(key) && key != parentEdgeLabel)
            {
                sb.AppendLine($"{indent}└─ {key}:");
                edgesIndent = indent + "   ";
            }

            // Enumerate and print the edges of the current node（entry.Key As edge tag）
            var entries = select.ToList();
            for (int i = 0; i < entries.Count; i++)
            {
                var entry = entries[i];
                bool isLastChild = i == entries.Count - 1;
                string connector = isLastChild ? "└─" : "├─";
                string label = entry.Key?.ToString() == Default ? "*" : entry.Key?.ToString();

                // If it is an optional parameter，Add(option)Identifier
                if (!string.IsNullOrEmpty(label) && optionalParams.Contains(label))
                {
                    label = label + "(option)";
                }

                if (entry.Value.func != null || entry.Value.contextFunc != null)
                {
                    string actionName;
                    if (entry.Value.contextFunc != null)
                    {
                        actionName = entry.Value.contextFunc.Method?.Name ?? "Anonymous";
                    }
                    else
                    {
                        actionName = entry.Value.func.Method?.Name ?? "Anonymous";
                    }
                    sb.AppendLine($"{edgesIndent}{connector} {label} → {actionName}");
                }
                else
                {
                    // Print edge tag
                    sb.AppendLine($"{edgesIndent}{connector} {label}");
                    // Recursively to child nodes；If the child node's key Different from the edge tag，Then print it in the subhierarchy key
                    string nextIndent = edgesIndent + (isLastChild ? "   " : "│  ");
                    entry.Value.Print(sb, nextIndent, isLastChild, label);
                }
            }
        }
        /// <summary>
        /// OverrideToStringMethod，Used to print state tree
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            var sb = new StringBuilder();
            Print(sb);
            return sb.ToString();
        }
    }
}

File: service/internal/statetree/statetree.go
package statetree

import (
	"encoding/json"
	"fmt"
	"strings"
)

const (
	DefaultKey = "*"
)

type ContextFunc func(ctx *Context) interface{}
type JsonFunc func(data map[string]interface{}) interface{}

type StateTree struct {
	Key            string
	Select         map[interface{}]*StateTree
	OptionalParams map[string]bool
	ContextFunc    ContextFunc
	JsonFunc       JsonFunc
	ErrorMessage   string
}

func New() *StateTree {
	return &StateTree{
		Select:         make(map[interface{}]*StateTree),
		OptionalParams: make(map[string]bool),
	}
}

func (st *StateTree) Run(ctx *Context) interface{} {
	current := st

	for current.ContextFunc == nil && current.JsonFunc == nil {
		keyToLookup := interface{}(DefaultKey)
		var next *StateTree

		if current.Key != "" && ctx != nil {
			if token, exists := ctx.GetJsonValue(current.Key); exists {
				keyToLookup = convertTokenToKey(token)
				if n, ok := current.Select[keyToLookup]; ok {
					next = n
				}
			}
		}

		if next == nil && ctx != nil {
			for k, v := range current.Select {
				if k == nil {
					continue
				}

				key := fmt.Sprintf("%v", k)
				if current.OptionalParams[key] {
					if paramToken, exists := ctx.GetJsonValue(key); exists && paramToken != nil {
						if !isNil(paramToken) && !isEmptyString(paramToken) {
							next = v
							break
						}
					}
				}
			}

			if next == nil {
				if n, ok := current.Select[DefaultKey]; ok {
					next = n
				}
			}
		}

		if next == nil {
			supportedKeys := []string{}
			for k := range current.Select {
				keyStr := fmt.Sprintf("%v", k)
				if keyStr != DefaultKey && !current.OptionalParams[keyStr] {
					supportedKeys = append(supportedKeys, keyStr)
				}
			}

			for k := range current.Select {
				keyStr := fmt.Sprintf("%v", k)
				if current.OptionalParams[keyStr] {
					supportedKeys = append(supportedKeys, keyStr+" (optional)")
				}
			}

			supportedKeysList := "none"
			if len(supportedKeys) > 0 {
				supportedKeysList = strings.Join(supportedKeys, ", ")
			}

			current.ErrorMessage = fmt.Sprintf("Invalid value '%v' for key '%s'. Supported values: [%s]", keyToLookup, current.Key, supportedKeysList)
			return nil
		}

		current = next
	}

	if current.ContextFunc != nil {
		return current.ContextFunc(ctx)
	}

	if current.JsonFunc != nil {
		return current.JsonFunc(ctx.JsonData)
	}

	return nil
}

func convertTokenToKey(token interface{}) interface{} {
	if token == nil {
		return DefaultKey
	}

	switch v := token.(type) {
	case float64:
		if v == float64(int64(v)) {
			return int64(v)
		}
		return v
	case bool:
		return v
	case string:
		if v == "" {
			return DefaultKey
		}
		return v
	case map[string]interface{}:
		if data, ok := v["Value"].(string); ok && data != "" {
			return data
		}
		return v
	default:
		return v
	}
}

func isNil(v interface{}) bool {
	if v == nil {
		return true
	}

	switch v.(type) {
	case json.Number:
		return false
	case map[string]interface{}:
		return false
	case []interface{}:
		return false
	default:
		return v == nil
	}
}

func isEmptyString(v interface{}) bool {
	if str, ok := v.(string); ok {
		return str == ""
	}
	return false
}

func (st *StateTree) String() string {
	var sb strings.Builder
	st.print(&sb, "", true, "")
	return sb.String()
}

func (st *StateTree) print(sb *strings.Builder, indent string, last bool, parentEdgeLabel string) {
	if indent == "" {
		sb.WriteString("StateTree\n")
	}

	edgesIndent := indent
	if st.Key != "" && st.Key != parentEdgeLabel {
		sb.WriteString(fmt.Sprintf("%s└─ %s:\n", indent, st.Key))
		edgesIndent = indent + "   "
	}

	entries := make([]struct {
		key   interface{}
		value *StateTree
	}, 0, len(st.Select))

	for k, v := range st.Select {
		entries = append(entries, struct {
			key   interface{}
			value *StateTree
		}{k, v})
	}

	for i, entry := range entries {
		isLastChild := i == len(entries)-1
		connector := "└─"
		if !isLastChild {
			connector = "├─"
		}

		label := fmt.Sprintf("%v", entry.key)
		if label == DefaultKey {
			label = "*"
		}

		if st.OptionalParams[label] {
			label = label + "(option)"
		}

		if entry.value.ContextFunc != nil || entry.value.JsonFunc != nil {
			actionName := "Anonymous"
			if entry.value.ContextFunc != nil {
				actionName = "ContextFunc"
			} else if entry.value.JsonFunc != nil {
				actionName = "JsonFunc"
			}
			sb.WriteString(fmt.Sprintf("%s%s %s → %s\n", edgesIndent, connector, label, actionName))
		} else {
			sb.WriteString(fmt.Sprintf("%s%s %s\n", edgesIndent, connector, label))
			nextIndent := edgesIndent
			if isLastChild {
				nextIndent = edgesIndent + "   "
			} else {
				nextIndent = edgesIndent + "│  "
			}
			entry.value.print(sb, nextIndent, isLastChild, label)
		}
	}
}
